xonsh/environ.py

```diff
@@ -631,20 +632,60 @@ def _repr_pretty_(self, p, cycle):
                 p.pretty(dict(self))
 
 
-def locate_binary(name, cwd):
-    # StackOverflow for `where` tip: http://stackoverflow.com/a/304447/90297
-    locator = 'where' if ON_WINDOWS else 'which'
-    try:
-        binary_location = subprocess.check_output([locator, name],
-                                                  cwd=cwd,
-                                                  stderr=subprocess.PIPE,
-                                                  universal_newlines=True)
-        if not binary_location:
-            return
-    except (subprocess.CalledProcessError, FileNotFoundError):
-        return
+def _is_executable_file(path):
+    """Checks that path is an executable regular file, or a symlink towards one.
+    This is roughly ``os.path isfile(path) and os.access(path, os.X_OK)``.
+
+    This function was forked from pexpect originally:
+
+    Copyright (c) 2013-2014, Pexpect development team
+    Copyright (c) 2012, Noah Spurrier <noah@noah.org>
+
+    PERMISSION TO USE, COPY, MODIFY, AND/OR DISTRIBUTE THIS SOFTWARE FOR ANY
+    PURPOSE WITH OR WITHOUT FEE IS HEREBY GRANTED, PROVIDED THAT THE ABOVE
+    COPYRIGHT NOTICE AND THIS PERMISSION NOTICE APPEAR IN ALL COPIES.
+    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+    """
+    # follow symlinks,
+    fpath = os.path.realpath(path)
+
+    if not os.path.isfile(fpath):
+        # non-files (directories, fifo, etc.)
+        return False
 
-    return binary_location
+    return os.access(fpath, os.X_OK)
+
+
+if ON_WINDOWS:
+    def yield_executables(directory, name):
+        extensions = builtins.__xonsh_env__.get('PATHEXT')
+        for a_file in os.listdir(directory):
+            base_name, ext = os.path.splitext(a_file)
+            if name == base_name and ext.upper() in extensions:
+                yield os.path.join(directory, a_file)
+else:
+    def yield_executables(directory, name):
+        if name in os.listdir(directory):
+            path = os.path.join(directory, name)
+            if _is_executable_file(path):
+                yield path
```

Comment:
Hmm, this is a little strange to me.  Why don't you have two function: `yield_executables_win()` and `yield_executables_posix()`, and then a top-level wrapper function called `yield_executables()` that has the if-else statement.  I think that is a bit cleaner over all.